### PR TITLE
Android - Support sending json arrays to server.

### PR DESCRIPTION
**Pull request by milanc**
Hi and thank you for creating this plugin.

While using it I've noticed that posting arrays is failing.
I am not a java developer and these are first lines of code I've written for java so if my PR is not good enough it might at least give some ideas what could be added to the current implementation.

Beside posting issues, I had issue with receiving plain values (e.g. boolean).
This PR has possible fix for this too.

BTW ios implementation has the same issue, but I don't have tools to try fix that one.

### CODE REVIEW

**Comment by ProofOfKeags** [comment]
For what it's worth, the JSON specification only allows Arrays and Objects as *top level values*. So receiving plain values (like `boolean`) should probably not be implemented as it is unsupported by the standard.

Additionally, this is also a problem in the ios side. I would recommend making the changes in tandem so that behavior is consistent across platforms.

Note: I am not a project maintainer, but we do have a fork of this plugin that we have augmented various ways and this is one of the things we did have to fix.

File: android/src/main/java/com/getcapacitor/plugin/http/Http.java [hunk 1]
```diff
@@ -453,17 +460,31 @@ private URL setParams(URL url, JSObject params) {
...
         DataOutputStream os = new DataOutputStream(conn.getOutputStream());
-        os.writeBytes(data.toString());
+
+        JSObject jsObject;
+        JSArray jsArray = null;
+        jsObject = call.getObject("data", null);
+        if (jsObject != null) {
+          os.writeBytes(jsObject.toString());
+        } else {
+          jsArray = call.getArray("data", null);
```
Comment (mlynch) [inline android/src/main/java/com/getcapacitor/plugin/http/Http.java:476]: Can you find a way to not mix these two conditionals together. Maybe handle the jsObject case completely and the jsArray case completely without mixing these `if`/`else`'s which is a bit hard to follow.

File: android/src/main/java/com/getcapacitor/plugin/http/Http.java [hunk 2]
```diff
@@ -391,8 +390,12 @@ private void buildResponse(PluginCall call, HttpURLConnection conn) throws Excep
           JSObject jsonValue = new JSObject(builder.toString());
           ret.put("data", jsonValue);
         } catch (JSONException e) {
-          JSArray jsonValue = new JSArray(builder.toString());
-          ret.put("data", jsonValue);
+            try {
+                JSArray jsonValue = new JSArray(builder.toString());
```
Comment (mlynch) [inline android/src/main/java/com/getcapacitor/plugin/http/Http.java:394]: Sorry, didn't have the prettier config set up, but can you change anywhere you're using four spaces to use two spaces (no tabs). This will be automated in the future, thanks.

File: android/src/main/java/com/getcapacitor/plugin/http/Http.java [hunk 3]
```diff
@@ -391,8 +390,12 @@ private void buildResponse(PluginCall call, HttpURLConnection conn) throws Excep
...
+                ret.put("data", jsonValue);
+            } catch (JSONException e2){
```
Comment (mlynch) [inline android/src/main/java/com/getcapacitor/plugin/http/Http.java:396]: Need Space after `)` and before the last `{`

File: ios/Plugin/Plugin.swift [hunk 4]
```diff
@@ -340,19 +340,27 @@ public class CAPHttpPlugin: CAPPlugin {
...
   }
-  
-  func setRequestDataJson(_ request: URLRequest, _ data: [String:Any]) throws -> Data? {
-    let jsonData = try JSONSerialization.data(withJSONObject: data)
+
+  func setRequestDataJson(_ request: URLRequest, _ call: CAPPluginCall) throws -> Data? {
+    let jsonData: Data
+    let jsObject = call.getObject("data")
+
+    if (jsObject != nil){
```
Comment (mlynch) [inline ios/Plugin/Plugin.swift:358]: Space before last `{`

File: ios/Plugin/Plugin.swift [hunk 5]
```diff
@@ -340,19 +340,27 @@ public class CAPHttpPlugin: CAPPlugin {
...
+        jsonData = try JSONSerialization.data(withJSONObject: jsObject as Any)
+    } else {
+        let jsArray = call.getArray("data", (Any).self)
```
Comment (mlynch) [inline ios/Plugin/Plugin.swift:361]: Two spaces

**Comment by milanc** [comment]
Hi @mlynch 
Thanks for the review, will fix these tomorrow when I get on my work machine.
BTW I have fix for responses that have content type subtype like xyz+json, but I may not have solution for ios.
Are you ok with PRs just for one platform?

**Comment by sbannigan** [comment]
@mlynch @milanc any update on this? We just ran into this issue as well. I'd be glad to help out if there is any additional work to be done.

**Comment by milanc** [comment]
Hi, @sbannigan, I assume it just needs to be merged.

File: ios/Plugin/Plugin.swift [hunk 6]
```diff
@@ -340,19 +340,27 @@ public class CAPHttpPlugin: CAPPlugin {
...
-    let jsonData = try JSONSerialization.data(withJSONObject: data)
+
+  func setRequestDataJson(_ request: URLRequest, _ call: CAPPluginCall) throws -> Data? {
+    let jsonData: Data
+    let jsObject = call.getObject("data")
+
+    if (jsObject != nil) {
+      jsonData = try JSONSerialization.data(withJSONObject: jsObject as Any)
+    } else {
+      let jsArray = call.getArray("data", (Any).self)
```
Comment (sbannigan) [inline ios/Plugin/Plugin.swift:361]: Without a default value here, apps will crash when trying to `POST` with `null` in the body
```suggestion
      let jsArray = call.getArray("data", (Any).self, [])
```

Comment (milanc) [reply ios/Plugin/Plugin.swift]: Thank you for suggestion, it is committed.

**Comment by yacut** [comment]
Any update on this?

**Comment by thomasvidas** [comment]
Unfortunately, this PR is just too stale at this point. I will add this feature to the `capacitor-v3` branch and use this PR as reference.

**Comment by milanc** [comment]
Few days ago I got back to the project that is using this plugin and merged upstream.
I was planning to push those changes to this PR.
I can create new PR  if you want.

**Comment by thomasvidas** [comment]
If you'd can to please do! 😄
